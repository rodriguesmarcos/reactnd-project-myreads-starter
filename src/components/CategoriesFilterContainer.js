import React, { Component } from 'react';
import PropTypes from 'prop-types';
import sortBy from 'sort-by';
import { onlyUnique } from '../helpers';
import CategoriesFilter from './CategoriesFilter';

class CategoriesFilterContainer  extends Component {

  constructor() {
    super();

    this.generateCategoriesArray = this.generateCategoriesArray.bind(this);
    this.mapCategories = this.mapCategories.bind(this);
  }

  state = {
    categories: []
  }

  componentDidMount() {
    const { books } = this.props;
    if ( books.length > 0 ) {
      this.updateCategories(books);
    }
  }

  componentWillReceiveProps(nextProps) {
    const { books: nextShowing } = nextProps;
    const { books } = this.props

    if ( JSON.stringify(nextShowing) !== JSON.stringify(books)) {
      this.updateCategories(nextShowing);
    }
  }

  updateCategories(books) {
    const allCategories = this.generateCategoriesArray(books)
    const categories = this.mapCategories(allCategories)
      .sort( sortBy( 'name' ));;
    this.setState({ categories });
  }

  /**
  * @description Generate an array of categories from books
  * @param {Array} prevProps
  */
  generateCategoriesArray(books) {
    const allCategories = books.map( b => b.categories )
      .reduce( (arr, val) => arr.concat(val), [] );
    const categories = allCategories.filter(onlyUnique);
    return categories;
  }

  mapCategories(categories) {
    const catObjects = categories.map( c => ( { id: c, name: c } ) );
    return catObjects;
  }

  render() {
    const { currCat, onChangeCategory } = this.props;
    const { categories } = this.state;

    return (
      <CategoriesFilter
        categories={categories}
        currCat={currCat}
        onChangeCategory={onChangeCategory}
      />
    )
  }
}

CategoriesFilterContainer.propTypes = {
  books: PropTypes.array.isRequired,
  currCat: PropTypes.string.isRequired,
  onChangeCategory: PropTypes.func.isRequired,
}

export default CategoriesFilterContainer;
