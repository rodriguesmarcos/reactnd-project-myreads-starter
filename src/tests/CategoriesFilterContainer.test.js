import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import CategoriesFilterContainer from '../components/CategoriesFilterContainer.js';
import books from './fixtures/books';
import shelves from './fixtures/shelves';

describe('<CategoriesFilterContainer />', () => {

  const minProps = {
    books: books,
    currCat: '',
    onChangeCategory: () => {},
  }

  test('renders without exploding', () => {
    const wrapper = shallow(<CategoriesFilterContainer {...minProps}/>);
    expect(wrapper.length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

});
