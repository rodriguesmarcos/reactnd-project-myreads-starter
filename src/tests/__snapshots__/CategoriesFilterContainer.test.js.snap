exports[`<CategoriesFilterContainer /> renders without exploding 1`] = `
<CategoriesFilter
  categories={
    Array [
      Object {
        "id": "COMPUTERS",
        "name": "COMPUTERS",
      },
      Object {
        "id": undefined,
        "name": undefined,
      },
      Object {
        "id": "Business & Economics",
        "name": "Business & Economics",
      },
      Object {
        "id": "Computers",
        "name": "Computers",
      },
      Object {
        "id": "Fiction",
        "name": "Fiction",
      },
      Object {
        "id": "Performing Arts",
        "name": "Performing Arts",
      },
    ]
  }
  currCat=""
  onChangeCategory={[Function]} />
`;
